rust/src/encodings.rs

```diff
@@ -46,12 +47,12 @@ pub trait Encoder {
 }
 
 /// Decoder.
-pub trait Decoder<T: ArrowPrimitiveType> {
+pub trait Decoder<T: NativeType> {
     type ArrowType;
 
-    fn decode(&mut self, offset: i32, length: &Option<i32>) -> Result<ArrayRef>;
+    fn decode(&mut self, offset: i32, length: &Option<i32>) -> Result<Arc<dyn Array>> ;
```

Comment:
Why some of them return `Arc<dyn Array>` some of them return `Box<dyn Array>`

Reply:
The `Arc` is a mimic of the original arrow-rs version, the `Box` is deduced from [this](https://github.com/jorgecarleitao/arrow2/blob/1b7e8eefc6741f9ff7b050b56db32f7f895edd44/src/compute/take/mod.rs#L40) arrow2 function. It's very similar to the [arrow-rs](https://github.com/apache/arrow-rs/blob/77c814cfc66845192c1e0be6ae8b0de6a1a0d5f0/arrow/src/compute/kernels/take.rs#L94) version except for the signature. I don't know the design intention of Arrow2 here, just keep it unchanged when it works.

Reply:
Can we make our interface consistent? either all return `Box<dyn Array>` or all `<Arc<dyn Array>>`?
Conceptually, decode / take should all return "array" from user API's aspect.

Reply:
Switched to box